Store graph edges in SharedArrayBuffer

lettertwo opened the pull request:
# Store graph edges in SharedArrayBuffer

<!---
Provide a general summary of the pull request here
Please look for any issues that this PR resolves and tag them in the PR.
-->

This changes the low level storage mechanism for graph adjacency lists from multiple [JS Maps] to a custom hashmap (😅 ) implemented on top of [SharedArrayBuffer]. 

---
Background: Why would we want to implement our own hashmap?

## Parallelism

Parcel has the ability to run many operations in parallel, but Node does not have any way to share data between threads except by serializing data and sending it to *each worker*, where it is then deserialized *by each worker* before work commences. Even for modestly sized Parcel projects, this overhead is significant. 

JavaScript _does_ have support for sharing memory between threads via [SharedArrayBuffer], but unfortunately, there is currently nothing higher level (like a `SharedMap` or something) in JS or Node that can take advantage of this in a significant way.

Thus, having a custom data structure built on top of `SharedArrayBuffer` means that Parcel can build a very large graph and share some significant portion of that graph with workers *for free*.

## Caching

Parcel caches build state frequently, and that involves serializing data to be written to cache. an `ArrayBuffer` of data serializes more quickly than a `Map`. 

## Other Solutions Considered

We evaluated other options, such as [Protocol Buffers] and [FlatBuffers] (either of which may still be viable for storing node data in the future), but ruled them out for this work for the following reasons:
- Any transport format still requires serialization and deserialization, which is inherently less efficient than sharing memory
-  These solutions generally incur a build step to generate bindings for reading and writing data to the underlying transport format
- The adjacency list only stores integers (node ids, edge types) which makes it good candidate for storage  in an `ArrayBuffer`

---

We've added a small set of unit tests to exercise the basic assumptions about our implementation, but we are mostly relying on the existing suites of tests to verify this work.

We have also done extensive benchmarking to measure the impact of these changes in a real-world, significantly sized app.

---

> **EDIT:** Refactoring based on feedback seems to have improved the overall performance enough to eliminate most of the regressions!


Benchmarks: Testing the impact of these changes in a real app

---

These benchmarks were taken on a Parcel app with a bundle graph containing **~92,421 nodes** and **~255,538 edges** and a request graph containing **~163,633 nodes** and **~1,351,038 edges**.

Some definitions:

<dl>
<dt>Cold Build</dt>
<dd>A parcel production build with no cache.</dd>
<dt>Warm Build</dt>
<dd>A parcel production build from cache.</dd>
<dt>Cold Start</dt>
<dd>A parcel dev start with no cache.</dd>
<dt>Warm Start</dt>
<dd>A parcel dev start from cache.</dd>
<dt>Total {Serialization,Deserialization} Time</dt>
<dd>How much time is spent serializing or deserializing the graphs (including node data)</dd>
<dt>Edge {Serialization,Deserialization} Time</dt>
<dd>How much time is spent serializing or deserializing the adjacency list (just the edge data)</dd>
</dl>


| Cold Build                 |         v2 |    this PR |      change | change (%) |
| -------------------------- | ---------: | ---------: | ----------: | ---------: |
| Total Memory               | 3397.57 mb | 1755.89 mb | -1641.68 mb |   48% less |
| Total Serialization Time   |    4941 ms |    3309 ms |    -1632 ms | 33% faster |
| Edge Serialization Time    |    2372 ms |    1456 ms |     -916 ms | 39% faster |
| Total Deserialization Time |    5738 ms |    4123 ms |    -1615 ms | 28% faster |
| Edge Deserialization Time  |    2994 ms |     518 ms |    -2476 ms | 83% faster |
| init                       |       1 ms |       1 ms |       +0 ms | 16% slower |
| build                      |  163023 ms |  135714 ms |   -27309 ms | 17% faster |
| shutdown                   |    2991 ms |    4517 ms |    +1525 ms | 51% slower |
| ✨ Built in                |   146.39 s |   135.69 s |    -10.70 s |  7% faster |


| Warm Build                 |         v2 |   this PR |     change | change (%) |
| -------------------------- | ---------: | --------: | ---------: | ---------: |
| Total Memory               | 1304.12 mb | 946.77 mb | -357.35 mb |   27% less |
| Total Serialization Time   |    6469 ms |   5379 ms |   -1090 ms | 17% faster |
| Edge Serialization Time    |    3109 ms |   1438 ms |   -1671 ms | 54% faster |
| Total Deserialization Time |    6223 ms |   4563 ms |   -1660 ms | 27% faster |
| Edge Deserialization Time  |    3028 ms |    495 ms |   -2533 ms | 84% faster |
| init                       |    1994 ms |    744 ms |   -1250 ms | 63% faster |
| build                      |    4007 ms |   3304 ms |    -703 ms | 18% faster |
| shutdown                   |    2678 ms |   3901 ms |   +1223 ms | 46% slower |
| ✨ Built in                |     4.95 s |    3.79 s |    -1.16 s | 23% faster |


| Cold Start                 |         v2 |    this PR |     change | change (%) |
| -------------------------- | ---------: | ---------: | ---------: | ---------: |
| Total Memory               | 2337.65 mb | 1384.07 mb | -953.58 mb |   41% less |
| Total Serialization Time   |    4110 ms |    3049 ms |   -1061 ms | 26% faster |
| Edge Serialization Time    |    2358 ms |    1329 ms |   -1029 ms | 44% faster |
| Total Deserialization Time |    6441 ms |    3311 ms |   -3130 ms | 49% faster |
| Edge Deserialization Time  |    3141 ms |     356 ms |   -2785 ms | 89% faster |
| init                       |       1 ms |       2 ms |      +1 ms | 41% slower |
| build                      |   79695 ms |   69485 ms |  -10211 ms | 13% faster |
| save                       |     628 ms |     415 ms |    -213 ms | 34% faster |
| shutdown                   |    2565 ms |    3490 ms |    +925 ms | 36% slower |
| ✨ Built in                |    76.64 s |    69.47 s |    -7.17 s | 19% faster |


| Warm Start                 |         v2 |   this PR |     change | change (%) |
| -------------------------- | ---------: | --------: | ---------: | ---------: |
| Total Memory               | 1195.51 mb | 807.47 mb | -388.04 mb |   32% less |
| Total Serialization Time   |    6756 ms |   4198 ms |   -2558 ms | 38% faster |
| Edge Serialization Time    |    2326 ms |   1498 ms |    -828 ms | 36% faster |
| Total Deserialization Time |    6254 ms |   4073 ms |   -2181 ms | 35% faster |
| Edge Deserialization Time  |    2897 ms |    501 ms |   -2396 ms | 83% faster |
| init                       |    1934 ms |    907 ms |   -1028 ms | 53% faster |
| build                      |    3382 ms |   2418 ms |    -964 ms | 29% faster |
| save                       |     999 ms |    464 ms |    -535 ms | 54% faster |
| shutdown                   |    2564 ms |   3067 ms |    +503 ms | 20% slower |
| ✨ Built in                |     2.57 s |    2.42 s |    -0.15 s |  6% faster |

---

## Summary

- Total memory usage is down significantly across all tested conditions
- Serialization and deserialization are both significantly faster in all tested conditions
- ~~With warm cache,~~ builds are on par **or faster** in all tested conditions
- Shutdown times seem to be slower, which we did not expect, and requires further investigation.

~~While we have seen some regression with cold build times, and not much improvement with warm times,~~ We have seen no regression (and even some improvement!) in build times, and we have also seen significant improvement in memory usage and serialization and deserialization times in all build scenarios.  However, shutdown times are slower than expected.

The main reasons we think these are acceptable trade-offs are:

- We believe we have not significantly regressed the most common DX scenarios
- These results reflect one part of the graph improvements we are working toward. Our hope and expectation is that we will ultimately improve all of these metrics in future by backing the graph nodes in shared buffers like we have done for the edges here.

For this work, we did not address data stored in the graph nodes, focusing solely on the edges. Since there are generally many more edges than nodes in most of Parcel's graphs, this change represents a measurable improvement, even though there is still more to do to improve the efficiency of the graph structure. 

[JS Maps]: https://github.com/parcel-bundler/parcel/blob/fb9818a/packages/core/graph/src/Graph.js#L573-L623
[SharedArrayBuffer]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/SharedArrayBuffer
[Protocol Buffers]: https://developers.google.com/protocol-buffers
[FlatBuffers]: https://google.github.io/flatbuffers/

thebriando committed:
change edgetypes to numbers

thebriando committed:
Merge remote-tracking branch 'origin/v2' into bdo/number-edgetypes

thebriando committed:
clean up comments

thebriando committed:
fix edge colors in graphviz

thebriando committed:
camelCase edge type objects

thebriando committed:
add NullEdgeType to generic in Array edge types

thebriando committed:
Merge branch 'v2' into bdo/number-edgetypes

thebriando committed:
Merge remote-tracking branch 'origin/v2' into bdo/number-edgetypes

lettertwo committed:
WIP: add new graph impl

lettertwo committed:
Update EfficientGraph plan

thebriando committed:
add support for edge types in getNodesConnectedTo/From, index, hasEdge

thebriando committed:
naive getAllEdges implementation

thebriando committed:
support array of edge types in getNodesConnectedFrom/to

thebriando committed:
fix behavior for getting nodes for all edge types

lettertwo committed:
Annotate resizeEdges

lettertwo committed:
Add graphviz() method

This is not likely to stay here; it is probably better integrated
into some test or benchmarking utilities (like `dumpGraphToGraphviz`).

lettertwo committed:
Add openGraphViz util

lettertwo committed:
Add 1 to edge hashes and type values

This is to allow truthy tests to pass when checking to see if
edges exist.

lettertwo committed:
Initial removeEdge implementation

lettertwo committed:
WIP: Use new EfficientGraph in Graph

thebriando committed:
use EfficientGraph getNodes functions, update graph tests, change default edge type to 1

thebriando committed:
Merge remote-tracking branch 'origin/lettertwo/buffer-backed-graph' into bdo/buffer-backed-graph

thebriando committed:
Use new EfficientGraph functions in Graph, update tests, fix edge type in removeEdge

thebriando committed:
fix traversal test

thebriando committed:
change NullEdgeType back to 1, enforce edge types to be non-zero in Graph

thebriando committed:
Add generic TEdgeType to EfficientGraph

thebriando committed:
Add generic TEdgeType to EfficientGraph

lettertwo committed:
Merge branch 'bdo/buffer-backed-graph' into lettertwo/buffer-backed-graph

* bdo/buffer-backed-graph:
  Add generic TEdgeType to EfficientGraph
  change NullEdgeType back to 1, enforce edge types to be non-zero in Graph
  fix traversal test
  Use new EfficientGraph functions in Graph, update tests, fix edge type in removeEdge
  use EfficientGraph getNodes functions, update graph tests, change default edge type to 1

lettertwo committed:
Use indexOfEdge and edgeAt helpers

lettertwo committed:
Fix node storage overlap

This adds `nodeAt` and `indexOfNode` utils to abstract away
the bookkeeping for storing and retrieving nodes in the byte array.

Also fixes a bug where subsequently added nodes would partially
overlap previous nodes in the byte array.

lettertwo committed:
Fix graphviz for nodes with multiple edges

lettertwo committed:
Fix removeEdge for nodes with multiple edges

lettertwo committed:
Fix isOrphanedNode

lettertwo committed:
Fix type mismatches between Graph and EfficientGraph

thebriando committed:
hash edge types

thebriando committed:
add getEdgesByType functions

lettertwo committed:
Fix Graph tests that check for null edge type

thebriando committed:
Merge remote-tracking branch 'origin/lettertwo/buffer-backed-graph' into bdo/buffer-backed-graph

lettertwo committed:
preserve edge insertion order in getAllEdges

Lots of test cases seem to expect this. I'm not sure if that is
intentional or not, but it seems reasonable to do this, regardless.

lettertwo committed:
Fix removeEdge for edges that have hash collisions

Also attempt to disambiguate some of the naming
for hash and index functions.

thebriando committed:
Remove old AdjacencyList, serialize EfficientGraph in Graph, update BundleGraph to use new functions

thebriando committed:
Merge remote-tracking branch 'origin/v2' into bdo/buffer-backed-graph

lettertwo committed:
Merge branch 'bdo/buffer-backed-graph' into lettertwo/buffer-backed-graph

* bdo/buffer-backed-graph: (37 commits)
  Improve resolver performance (#6328)
  v2.0.0-beta.3.1
  Add it to the nightly release workflow as well
  Configure jemalloc page size for M1 (#6314)
  Update swc (#6307)
  Fix parcelDependencies
  v2.0.0-beta.3
  Update swc (#6289)
  Remove old AdjacencyList, serialize EfficientGraph in Graph, update BundleGraph to use new functions
  Babel ast location, diagnostic, and source location remapping (#6238)
  update contributing guide (#6293)
  Update sourcemap to rc-1.0 (#6279)
  Serve nearest index.html in case the applications has multiple index files (#6250)
  Allow bundling browserify bundles by replacing free requires with undefined (#6260)
  Upgrade flow to 0.151.0 (#6281)
  Use local require for `@babel/core` and `postcss` (#6264)
  Resolve helpers relative to JS transformer (#6278)
  Pure comment for $parcel$interopDefault (#6271)
  Don't transpile spread in JSX with modern targets (#6274)
  Update TS validator assertions (#6272)
  ...

lettertwo committed:
Clean up EfficientGraph tests

lettertwo committed:
Add removeNode method

devongovett committed:
Merge branch 'v2' of github.com:parcel-bundler/parcel into bdo/number-edgetypes

# Conflicts:
#	packages/core/core/src/BundleGraph.js
#	packages/core/core/src/dumpGraphToGraphViz.js
#	yarn.lock

lettertwo committed:
Remove some old comments

lettertwo committed:
Fix resize loop when initial graph size is very small

We resize the graph edges when the available capacity is <30%,
but we were indexing a new edge _before_ expanding capacity,
which was causing a loop when the the total capacity was too small
(i.e., `4`) to trigger the 30% threshold in a previous add _and_
there was no available capcity for the next add.

lettertwo committed:
Use `indexOf` when checking if graph as an edge

lettertwo committed:
Fix resizeEdges not updating incoming and outgoing references

lettertwo committed:
Add simple clustering and uniformity stats

This could help us optimize our hash function.

lettertwo committed:
Hash based on order of arguments and modulo EDGE_SIZE

lettertwo committed:
Deserialize EfficientGraph from provided serialized data

lettertwo committed:
Rehash edges when node capacity changes

lettertwo committed:
Serialize node and edge capacities

Since these values are used in hashing, we serialize them
along with the edges and nodes to ensure that the hashes don't
change when a new graph is deserialized from a previous graph.

lettertwo committed:
Fix edge iterations when edge index is 0

lettertwo committed:
Fix typos and make clarifications

lettertwo committed:
Fix lookups after edges have been removed

This addes a DELETED sentinel that preserves contiguity
in the hash table so that scanning forward to find edges
that match a hash continues to work after other edges with
a matching hash have been removed.

lettertwo committed:
Add tests for addEdge, removeEdge

thebriando committed:
Fix incoming/outgoing edges in removeEdge

thebriando committed:
change assert.equal to assert.deepEqual

lettertwo committed:
Remove unused removeNode method

Not only is it unused, but it might not make any sense for it to exist,
given that adding a node is basically just incrementing a counter.

lettertwo committed:
Fix typo

lettertwo committed:
Use stored node/edge capacity instead of recomputing

Also fix collision stat and add node/edge capacity and load stats

thebriando committed:
Merge remote-tracking branch 'origin/v2' into lettertwo/buffer-backed-graph

thebriando committed:
fix hash function

lettertwo committed:
Fix edge graphviz

thebriando committed:
Revert "fix hash function"

This reverts commit 82bb1753c5a5ed6c2674d5b983f104e0ea00265a.

thebriando committed:
Fix inconsistencies with index/hash values in removeEdge/addEdge

lettertwo committed:
Deduplicate nodes in getNodesConnected{To,From}

thebriando committed:
Fix indexFor returning indexes for edges that already exist

lettertwo committed:
Rename EfficientGraph -> AdjacencyList

lettertwo committed:
Make getAllEdges iterable

The common use case here is iterating over all the edges
to copy them to a new graph. Since a graph may have _a lot_
of edges, we can conserve memory by avoiding collecting all
edges up front.

lettertwo committed:
Bail earlier in isOrphan checks

lettertwo committed:
Fix flow errors

lettertwo committed:
Clean up Graph changes

lettertwo committed:
Nit: Remove redundant array spread

lettertwo committed:
Update addEdge tests for managing deleted edges

thebriando committed:
remove nextNodeId

thebriando committed:
Add LAST_IN and LAST_OUT references to node array for constant time insertion

thebriando committed:
Fix references in addEdge/removeEdge

lettertwo committed:
Refactor AdjacencyList to Node and Edge views

These views are meant to simplify the bookkeeping operations
in AdjacencyList, though they may be introducing some slowdowns.

However, this should provide some optimization opportunity,
for example, caching/memoizing of computed values,
and also implementing a doubly linked edge list to improve
scan / iteration times.

lettertwo committed:
Add inspection methods to Node, Edge

This makes console logging a little more readable during debugging

lettertwo committed:
Move edge linking out of Edge statics

We'll try to make AdjacencyList be responsible for all linking,
while Edge/Node are just responsible for reading/writing to/from
the node and edge arrays.

thebriando committed:
Add prevIn/prevOut references for constant time removal in removeEdge

lettertwo committed:
Update indexOf/indexFor to default to null edge type

Aslo remove some old code

lettertwo committed:
Simplify resizing; add Edge and Node tests

lettertwo committed:
Merge branch 'v2' into lettertwo/buffer-backed-graph

* v2: (34 commits)
  Wrap assets recursively when any incoming dependency is wrapped (#6572)
  Improvements for library targets (#6570)
  Diagnostic for undeclared external dependencies in library builds (#6564)
  More bugs (#6567)
  Don't require `url:` for image transformer (#6565)
  Remove 'Name already registered with serializer' error (#6566)
  Fix live bindings and `this` of external CommonJS modules (#6548)
  JS runtime improvements (#6531)
  Make sure the absolute path isn't contained in the cache (#5900)
  Adds '@parcel/diagnostic' to dependencies (#6563)
  Disable workers with string literals and improve diagnostics (#6536)
  Bug fixes (#6541)
  Don't attempt to resolve URLs starting with '#' (#6504)
  Correctly set worker's output format if not support by environment (#6534)
  Babel: Recognize peerDependencies in isJSX (#6497)
  fix setHeaders ordering on dev server (#6500)
  Graph: Remove Node interface (#6530)
  Fix TS build script for old Node versions (#6526)
  Improve library targets (#6517)
  Fix TypeScript and other sourcemaps by always creating an initial sourcemap (#6472)
  ...

lettertwo committed:
Speed up traversal to/from a node by edge type.

This adds a lookup table by type to or from a given node,
so that it is trivial to frequently get a given node's
connected nodes of a given type.

lettertwo committed:
Use type map when generating typed edges

lettertwo committed:
Merge Edge and Node into AdjacencyList

The overhead associated with these two data structures
was not negligible, so a better choice is to use methods
on AdjacencyList instead of creating Edge and Node views.

lettertwo committed:
Add maxCollisions stat

lettertwo committed:
Remove old tests

lettertwo committed:
Make getAllEdges() preserve insertion order

lettertwo committed:
Implement XOR doubly linked edge list

This is a space optimization; there may be some performance impact
due to the extra time spent scanning the list when removing an edge.

It saves space by eliminating the need to explicitly store the
next and previous edges in each edge object. Instead, it maintains
an XOR of the next and previous edges on each edge, which means
following the links between edges requires knowing the edge that
came before or after.

lettertwo committed:
Microoptimizing

lettertwo committed:
Refactor generator methods to return arrays

They're all just getting immediately exhausted by Graph anyway.

lettertwo committed:
Revert to singly linked edge list

XOR links are about 25% slower than the reference link implementation,
but use about 20-25% less space. The singly linked version is even
slower than XOR, but has the same space savings.

This commit restores a singly linked list, but adds a caching
mechanism for improving the perf when finding a previous edge.

lettertwo committed:
Remove old debugging code

lettertwo committed:
Make deserialized AdjacencyList readonly; add clone method

thebriando committed:
Memoize edge indexes

lettertwo committed:
Implement a close addressed hash table

Also added a header sequence to the arrays, which eliminates
the need to serialize other data separately from the arrays.

thebriando committed:
Bookkeep empty/deleted edges for constant time lookup when we search for an available slot to add an edge to

thebriando committed:
Refactor hashing and index functions to use coalesced chaining

Each edge now has a pointer to the next edge in the list of collisions
indexFor will find an empty edge and update the last edge in the chain to point to the newly added edge
indexOf will now only traverse through the list of edges with the same hash

TODO:
* reuse deleted edges
* move NEXT_HASH pointer to another structure to save on memory
* the cellar gets exhausted pretty quickly, look into increasing the amount of space the cellar occupies

thebriando committed:
Merge remote-tracking branch 'origin/v2' into bdo/number-edgetypes

thebriando committed:
Change NullEdgeType to 1

thebriando committed:
Reuse deleted edges when searching for usable edges

thebriando committed:
Keep NEXT_HASH reference when removing edges in the adressable space that point to edges in the cellar

lettertwo committed:
Fix unit tests

lettertwo committed:
Defer reclaiming of deleted space until the next add

We were reclaiming space when removing edges, but it turns out
that doing so was quite expensive in practice. Instead, we defer
reclaiming that space until the sum total of adds and deletes
exceeds the capacity by the load factor.

lettertwo committed:
(Re)introduce doubly linked adjacency list

lettertwo committed:
Remove DELETED sentinel

We don't need this any more because we eliminated linear scanning.

lettertwo committed:
Update comments to reflect changes

lettertwo committed:
Build typemap cache incrementally and on demand

lettertwo committed:
Update comments to reflect changes

lettertwo committed:
Fix edge capacity stats; add buffer size stats

lettertwo committed:
Add min and max node and edge capacity checks

lettertwo committed:
Ease edge growth rate as capacity increases

This is intended to strike a balance between growing the
edge capacity in too small increments, which causes a lot of
resizing, and growing the edge capacity in too large increments,
which results in a lot of wasted memory.

lettertwo committed:
Lazily populate typemap cache

It turns out prepopulating the typemap before doing any traversal
slows down the graph building process significantly, but does not
make traversal any faster.

lettertwo committed:
Fix tests

lettertwo committed:
Merge branch 'v2' into bdo/number-edgetypes

* v2: (42 commits)
  Remove dead link to docs in diagnostic (#6913)
  Fix engines.parcel in SVG packager (#6911)
  Use imported Readable flow type instead of global (#6910)
  Use non-deprecated SVGO options in HTMLNanoOptimizer (#6785)
  Use yarnpkg registry for all the things (#6908)
  Fix issue about loading configs from tsconfig.json (#6881)
  Add `@section` to Compressor type (#6885)
  Parcel API improvements (#6866)
  Use stream-browserify for polyfilling instead (#6863)
  Add support for compressor plugins (#6776)
  Support for inline style attributes and inline scripts in SVG (#6797)
  Bump less (#6852)
  Make sure (non-React) SVGs are in separate bundles (#6757)
  Add support for injecting manifest into service workers (#6798)
  Fix 'does not export default' error with scope hoisting and url/worklet pipeline (#6803)
  Remaining cargo clippy fixes (#6829)
  Lazily install sharp only when needed (#6816)
  More cargo clippy suggestions (#6811)
  Don't emit runtime manifest for inline child bundles (#6807)
  Apply suggestions from rust-clippy (#6256)
  ...

lettertwo committed:
Merge branch 'bdo/number-edgetypes' into lettertwo/buffer-backed-graph

* bdo/number-edgetypes: (74 commits)
  Remove dead link to docs in diagnostic (#6913)
  Fix engines.parcel in SVG packager (#6911)
  Use imported Readable flow type instead of global (#6910)
  Use non-deprecated SVGO options in HTMLNanoOptimizer (#6785)
  Use yarnpkg registry for all the things (#6908)
  Fix issue about loading configs from tsconfig.json (#6881)
  Add `@section` to Compressor type (#6885)
  Parcel API improvements (#6866)
  Use stream-browserify for polyfilling instead (#6863)
  Add support for compressor plugins (#6776)
  Support for inline style attributes and inline scripts in SVG (#6797)
  Bump less (#6852)
  Make sure (non-React) SVGs are in separate bundles (#6757)
  Add support for injecting manifest into service workers (#6798)
  Fix 'does not export default' error with scope hoisting and url/worklet pipeline (#6803)
  Remaining cargo clippy fixes (#6829)
  Lazily install sharp only when needed (#6816)
  More cargo clippy suggestions (#6811)
  Don't emit runtime manifest for inline child bundles (#6807)
  Apply suggestions from rust-clippy (#6256)
  ...

lettertwo commented on packages/core/graph/src/AdjacencyList.js:
```diff
@@ -0,0 +1,1195 @@
+import type {NullEdgeType, AllEdgeTypes} from './Graph';
+import type {NodeId} from './types';
+
+/**
```
The implementation described here was inspired by [This dissection of the v8 Map internals](https://itnext.io/v8-deep-dives-understanding-map-internals-45eb94a183df)

lettertwo commented on packages/core/graph/src/AdjacencyList.js:
```diff
@@ -0,0 +1,1195 @@
+opaque type EdgeIndex = number;
+
+// From https://gist.github.com/badboy/6267743#32-bit-mix-functions
+function hash32shift(key: number): number {
```
This crazy bit mixing function improves the uniformity of our hash table a fair amount without adding any noticeable time to hashing. We found this in the [v8 Map hash utils](https://github.com/nodejs/node/blob/238104c531219db05e3421521c305404ce0c0cce/deps/v8/src/utils/utils.h#L213)

lettertwo commented on packages/core/graph/src/AdjacencyList.js:
```diff
@@ -0,0 +1,1195 @@
+   * Returns a serializable object of the nodes and edges in the graph.
+   */
+  serialize(): SerializedAdjacencyList<TEdgeType> {
+    return {
+      nodes: this.#nodes,
+      edges: this.#edges,
+    };
```
The improvements to serialization and deserialization times come here; `nodes` and `edges` are both `SharedArrayBuffer` instances.

thebriando commented on packages/core/graph/src/AdjacencyList.js:
```diff
@@ -0,0 +1,1195 @@
+  /**
+   * Get the list of nodes connected from this node.
+   */
+  getNodesConnectedFrom(
```
These functions can probably be renamed to `getNodeIdsConnected..` to match the ones in `Graph.js`.

parcel-benchmark commented:
## Benchmark Results
<details><summary>Kitchen Sink ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 1.95s | +39.00ms |
| Cached | 235.00ms | -26.00ms 🚀 |

#### Cold Bundles

*No bundle changes detected.*

#### Cached Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/legacy/index.96016b08.js | 1.59kb | +0.00b | 784.00ms | -41.00ms 🚀 |
| dist/legacy/index.8aaa89c9.js | 1.20kb | +0.00b | 785.00ms | -40.00ms 🚀 |
| dist/modern/index.6be20f01.js | 1.13kb | +0.00b | 784.00ms | -41.00ms 🚀 |

</p></details>

<details><summary>React HackerNews ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 9.37s | -20.00ms |
| Cached | 402.00ms | -41.00ms 🚀 |

#### Cold Bundles

*No bundle changes detected.*

#### Cached Bundles

*No bundle changes detected.*

</p></details>

<details><summary>AtlasKit Editor ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 1.02m | -83.00ms |
| Cached | 1.27s | -280.00ms 🚀 |

#### Cold Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/EmojiPickerComponent.07ae4435.js | 147.12kb | +0.00b | 18.94s | -9.59s 🚀 |
| dist/esm.f1a592e5.js | 33.15kb | +0.00b | 18.94s | -9.59s 🚀 |
| dist/DatePicker.f2994e45.js | 22.98kb | +0.00b | 18.94s | -9.59s 🚀 |
| dist/js.660b75db.js | 17.25kb | +0.00b | 34.16s | -11.87s 🚀 |
| dist/png-chunks-extract.5c9c0fb9.js | 3.58kb | +0.00b | 34.16s | -11.87s 🚀 |
| dist/feedback.c7e69064.js | 1.77kb | +0.00b | 18.93s | -9.59s 🚀 |
| dist/workerHasher.3639b1f9.js | 1.63kb | +0.00b | 18.94s | -9.59s 🚀 |
| dist/heading6.a693146e.js | 1.51kb | +0.00b | 18.93s | -9.59s 🚀 |
| dist/heading5.8bfafe21.js | 1.38kb | +0.00b | 18.93s | -9.59s 🚀 |
| dist/expand.a44d10fe.js | 1.29kb | +0.00b | 18.93s | -9.59s 🚀 |

#### Cached Bundles

*No bundle changes detected.*

</p></details>

<details><summary>Three.js ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 6.52s | +219.00ms |
| Cached | 318.00ms | -99.00ms 🚀 |

#### Cold Bundles

*No bundle changes detected.*

#### Cached Bundles

*No bundle changes detected.*

</p></details>

**[Click here to view a detailed benchmark overview.](https://benchmark.parceljs.org/benchmark/315376143717892674)**

devongovett commented on packages/core/graph/src/AdjacencyList.js:
```diff
@@ -0,0 +1,1195 @@
+    // .e.g., `hash(10, 24, 4) => 10244`.
+    // $FlowFixMe[unsafe-addition]
+    // $FlowFixMe[incompatible-type]
+    let hash = '' + from + to + type - 0;
```
This seems like it could be a perf issue. Converting numbers to strings and back is expensive and this is a very hot function.

Also, your hash32shift function only works on 32 bit integers, and this could easily create a number larger than that. Your node ids are up to 32 bits a piece already and you are concatenating them here.

Did you try some kind of hash combining function? No idea how good this on is but it might be faster?

```js
let hash = 17;
hash = hash * 37 + hash32shift(from);
hash = hash * 37 + hash32shift(to);
hash = hash * 37 + hash32shift(type);
```

lettertwo replied:
Very interesting! We thought string conversion seemed like a not great idea, but it was never indicated as a bottleneck in our profiling, so we left it. However, I've done some microbenching with your suggested impl, and it is indeed faster, but does not appear to degrade uniformity at all, which seems like a no-brainer:

'hash' is string conversion
'hash2' is prime multiplication

163,633 nodes, 1,351,038 edges:

|       | ops/sec | max collisions | avg collisions |
| :---- | ------: | -------------: | -------------: |
| hash  | 505,108 |              8 |           0.46 |
| hash2 | 884,441 |              9 |           0.46 |

92,421 nodes, 255,538 edges:

|       |   ops/sec | max collisions | avg collisions |
| :---- | --------: | -------------: | -------------: |
| hash  |   924,997 |              5 |           0.34 |
| hash2 | 1,384,921 |              7 |           0.34 |

devongovett commented on packages/core/graph/src/AdjacencyList.js:
```diff
@@ -0,0 +1,1195 @@
+  /** A map of node ids from => through types => to node ids. */
+  #from: Map<NodeId, Map<TEdgeType | NullEdgeType, Set<NodeId>>> = new Map();
+  /** A map of node ids to => through types => from node ids. */
+  #to: Map<NodeId, Map<TEdgeType | NullEdgeType, Set<NodeId>>> = new Map();
```
How much do these extra maps affect performance? We're basically paying the cost of 3 hash maps instead of just one, so basically the cost of our old implementation PLUS the cost of your new implementation (minus serializing these maps). Explains why it is slightly slower in some cases.

Would it make sense to adjust the main adjacency list data structure instead to be more oriented around iterating by type? It's pretty rare that we access all of the edges connected to a node across all edge types. Most of the time, we are asking for a single type. One idea could be to change the NEXT_IN, NEXT_OUT, etc. pointers to be the next edge of the same type rather than having a single linked list of every edge connected to a node. The nodes would gain a type in addition to FIRST_IN/FIRST_OUT etc., so basically there'd be a "node" per id + type. This way you'd basically have a separate linked list per edge type and could simply walk that to iterate the connected edges. If you wanted all edge types, you could instead pass a list of those (which we know upfront).

lettertwo replied:
We refactored the data structure just as you suggested; These maps are gone now! 🎉

devongovett commented on packages/core/graph/src/AdjacencyList.js:
```diff
@@ -0,0 +1,1195 @@
+    if (type <= 0) throw new Error(`Unsupported edge type ${0}`);
+
+    // The edge is already in the graph; do nothing.
+    if (this.hasEdge(from, to, type)) return false;
```
I think this results in hashing twice when the edge is not already in the graph (the common case). You could inline this check below after hashing maybe?

devongovett commented on packages/core/graph/src/AdjacencyList.js:
```diff
@@ -0,0 +1,1195 @@
+  }
+
+  /** Iterate over outgoing edge hashes from the given `nodeId` the `AdjacencyList`. */
+  *iterateOutgoingEdges(nodeId: NodeId): Iterator<EdgeIndex> {
```
How much do generators affect performance vs say using an array?

devongovett commented on packages/core/graph/src/AdjacencyList.js:
```diff
@@ -0,0 +1,1195 @@
+ * └────────────┴───────────┴───────────┘
+ *
+ * Each edge is represented with 8 4-byte chunks:
+ * The first 4 bytes are the edge type.
```
We probably don't need to allocate that much space for the type? 1 byte (max 256 types) would be more than enough. Though I guess that makes things a little more complex and you'd have to deal with alignment...

lettertwo replied:
This is a really good point, and something I expanded on a bit in [this comment](https://github.com/parcel-bundler/parcel/pull/6922#discussion_r718652525).

lettertwo committed:
Rename getNodesConnected{To,From} to getNodeIdsConnected{To,From}

This matches the naming of the corresponding methods on Graph.

lettertwo committed:
Reduce duplicate hashing in addEdge

lettertwo committed:
Replace string conversion with prime multiplication in hash

It's benchmarking about 67% faster without any degradation in uniformity.

Also fixed some tests that were assuming some things about hashing.

thebriando commented on packages/core/graph/src/AdjacencyList.js:
```diff
@@ -0,0 +1,1205 @@
+/** The size after which to grow the edge capacity by the minimum factor. */
+const PEAK_EDGE_CAPACITY = 2 ** 18;
+
+export const ALL_EDGE_TYPES: AllEdgeTypes = '@@all_edge_types';
```
nit: You can get rid of this declaration and import it from `Graph`

lettertwo committed:
Merge branch 'v2' into lettertwo/buffer-backed-graph

* v2:
  Upgrade Flow to 0.160.1 (#6964)
  Only use error overlay if there's a document (#6960)
  Don't fail when HTML tags are implied (#6752)
  Reorder resolveOptions() env priority (#6904)
  Change edge types to numbers (#6126)
  Bump swc (#6848)
  Print diagnostics for scope hoisting bailouts at verbose log level (#6918)

lettertwo committed:
Refactor nodes and edges into type maps

This looks like a big change because a lot of stuff moved around.
The edges are currently stored the same way they were before,
but the hash table implementation has been extracted for use
in the new node map.

The node map is where the real change is; instead of just linking
nodes to addresses in the edge map, it now stores multiple sets
of links per node, one for each type of edge.

This allows us to eliminate the need for a type map cache,
as we are now able to look up edges by type in constant time.

thebriando committed:
Use numeric value for ALL_EDGE_TYPES, remove $FlowFixMe's

thebriando committed:
Fix flow errors

lettertwo committed:
Explain weird Flow comment type

lettertwo committed:
Fix lint errors

lettertwo committed:
Merge branch 'v2' into lettertwo/buffer-backed-graph

lettertwo commented on packages/core/graph/src/AdjacencyList.js:
```diff
@@ -0,0 +1,1210 @@
+}
+
+/**
+ * `SharedTypeMap` is a hashmap of items,
```
The implementation described here was inspired by [This dissection of the v8 Map internals](https://itnext.io/v8-deep-dives-understanding-map-internals-45eb94a183df)

lettertwo commented on packages/core/graph/src/AdjacencyList.js:
```diff
@@ -0,0 +1,1210 @@
+}
+
+// From https://gist.github.com/badboy/6267743#32-bit-mix-functions
+function hash32shift(key: number): number {
```
This crazy bit mixing function improves the uniformity of our hash table a fair amount without adding any noticeable time to hashing. We found this in the [v8 Map hash utils](https://github.com/nodejs/node/blob/238104c531219db05e3421521c305404ce0c0cce/deps/v8/src/utils/utils.h#L213)

lettertwo commented on packages/core/graph/src/AdjacencyList.js:
```diff
@@ -0,0 +1,1210 @@
+    return true;
+  }
+
+  *getAllEdges(): Iterator<{|
```
This is the now the only scenario where we generate values for iteration. All other traversals are now using arrays.

lettertwo commented on packages/core/graph/src/Graph.js:
```diff
@@ -292,19 +221,6 @@ export default class Graph<TNode, TEdgeType: number = 1> {
     this.nodes.set(nodeId, node);
   }
 
-  replaceNode(
-    fromNodeId: NodeId,
-    toNodeId: NodeId,
-    type: TEdgeType | NullEdgeType = 1,
-  ): void {
-    this._assertHasNodeId(fromNodeId);
-    for (let parent of this.inboundEdges.getEdges(fromNodeId, type)) {
-      this.addEdge(parent, toNodeId, type);
-      this.removeEdge(parent, fromNodeId, type);
-    }
-    this.removeNode(fromNodeId);
-  }
-
```
This method appears to be unused.

lettertwo commented on packages/core/graph/src/AdjacencyList.js:
```diff
@@ -0,0 +1,1210 @@
+   * Returns the id of the added node.
+   */
+  addNode(): NodeId {
+    let id = this.#nodes.getId();
```
Note that we don't _actually_ add an entry to the node map until it is connected  to another node id in `addEdge`.

lettertwo commented on packages/core/graph/src/AdjacencyList.js:
```diff
@@ -0,0 +1,1210 @@
+        fromNode = this.#nodes.addressOf(from, type);
+      }
+    }
+    if (toNode === null) toNode = this.#nodes.add(to, type);
+    if (fromNode === null) fromNode = this.#nodes.add(from, type);
```
If this edge is the first _of this type_ for either of the nodes, we add a new entry to the node map.

lettertwo commented on packages/core/graph/src/AdjacencyList.js:
```diff
@@ -0,0 +1,1210 @@
+  }
+}
+
+/**
```
The implementation described here was inspired by [This dissection of the v8 Map internals](https://itnext.io/v8-deep-dives-understanding-map-internals-45eb94a183df)

lettertwo commented on packages/core/graph/src/AdjacencyList.js:
```diff
@@ -0,0 +1,1210 @@
+  /** The number of items to accommodate per hash bucket. */
+  static BUCKET_SIZE: number = 2;
+
+  data: Uint32Array;
```
There would be significant memory saved if we used `DataView` instead, which would allow us to store the `type` field in 1 byte (instead of 4), but it also would complicate the implementation significantly.

The savings would come from the following:
- each node adds an entry for every unique edge type it is connected to, so in the worst case (every node had at least one edge of every type), that's `n * t * 3` bytes saved, where `n` is the number of nodes and `t` is the number of unique edge types in the graph.
- each edge keeps its type in its entry, which would be `e * 3` bytes saved, where `e` is the number of edges in the graph.

Additionally, typed arrays are _technically_ not portable, since they use the originating system's [endianess](https://en.wikipedia.org/wiki/Endianness)

Some of the reasons we haven't implemented this yet:
- `DataView` is likely not as well optimized as type arrays
- memory allocation is more complex, since we cannot rely on the uniform item size of a typed array
- reads and writes must now be aware of endianness
- In practice, most systems are LE, so the non-portable nature of typed array is probably a non-issue
- memory savings from this implementation are already significant compared to v8 Maps

thebriando reviewed (commented):
https://github.com/parcel-bundler/parcel/blob/8aa9c25df02bcfa21d74c59ed4c4bd31a6caaafa/packages/core/core/src/BundleGraph.js#L1210-L1213

I wonder if creating a new function to do this would have any benefit since we can traverse backwards with the doubly linked list now.

lettertwo committed:
Merge branch 'v2' into lettertwo/buffer-backed-graph

* v2: (68 commits)
  Fix RangeError in `not export` error with other file type (#7295)
  Apply sourcemap in @parcel/transformer-typescript-tsc (#7287)
  Fix side effects glob matching (#7288)
  Fix changelog headings
  v2.0.1
  Changelog for v2.0.1
  Resolve GLSL relative to the importer, not the asset (#7263)
  fix: add @parcel/diagnostic as dependency of @parcel/transformer-typescript-types (#7248)
  Fixed missing "Parcel" export member in Module "@parcel/core" (#7250)
  Add script to sync engines with core version (#7207)
  Bump swc (#7216)
  Make Webpack loader detection regex dramatically faster (#7226)
  swc optimizer (#7212)
  Update esbuild in optimizer (#7233)
  Properly visit member expressions (#7228)
  Update to prettier 2 (#7209)
  Fix serve mode with target override and target source fields (#7187)
  Update package.json to include the repository (#7184)
  fix #6730: add transformer-raw as dependency of config-webextension (#7193)
  Log warning instead of crash if image optimizer fails (#7119)
  ...

lettertwo committed:
prettier

devongovett commented:
🥳 🎉
